game/ui/general/aequipscreen.cpp

```diff
@@ -1434,14 +1434,15 @@ bool AEquipScreen::tryPlaceItem(sp<Agent> agent, Vec2<int> slotPos, bool *insuff
 	if (canAdd)
```

Comment:
If the theory is true about disappearing equipment - how can canAdd be true? Surely that's only true if the slot is empty (so not reloading), or if it's in the ammo_types array (as checked at line 1375)?

Reply:
I'm going to post the tryPlaceItem method here easy reference, I think I have the logic mostly figured out for these errors:

https://github.com/OpenApoc/OpenApoc/blob/63ef6567a5eb49c006ac93859c3e6c21027039e9/game/ui/general/aequipscreen.cpp#L1369-L1466

Here is what I think is happening for two different issues. First, if a grenade or medkit is placed over a weapon, it will try to load it like a clip.

- 1372 - agent->canAddEquipment returns false, finds equipment under the dragged equipment at this position.
- 1379 - dragged equipment is not the same type as the ammo type of the weapon, continues to 1399.
- 1399 - checks for a nearby offset position to place the item, finds one. Sets canAdd to true and sets slotPos to new offset position.
- 1434 - since canAdd is true, checks for equipmentUnderCursor. Importantly, this was set *before* the offset was found. Since that is the case, it thinks there is still an item in the slot and moves to 1444.
- 1444 - Immediately loads the equipment into the weapon,  error is thrown.

In this case, it seems the equipmentUnderCursor variable should be checked again if an offset is found.

Second, if an item is dragged close to a slot but misses, and the item dissapears.

- 1372 - agent->canAddEquipment returns false, finds equipment under the dragged equipment at this position.
- equipmentUnderCursor is not a weapon, continues to 1399.
- 1399 - checks for a nearby offset position to place the item, finds one. Sets canAdd to true and sets slotPos to new offset position.
- 1434 - Since canAdd is true, checks for equipmentUnderCursor. Finds the left arm armor (in my testing in one specific slot). Moves to 1444 since equipmentUnderCursor was found.
- 1444 - Loads equipment as ammo, nothing happens since no weapon in slot.
- 1445 - checks the dragged equipment for ammo greater than 0, since a medkit has 0 ammo, continues to 1464.
- 1464 - Nullifies the dragged equipment, removing the item forever.

Again, it seems the equipmentUnderCursor needs to be rechecked. I did not look into this as deeply when I made my initial PR. I now see that my fix worked because it ignored the equipmentUnderCursor check that was using the original position.

Thoughts on this assessment?

Reply:
Spent some time going through this morning, and your assessment does appear to be correct
The fix does resolve a number of issues